manifests/instance.pp

```diff
@@ -0,0 +1,156 @@
+# = Define: nodejs::install
+#
+# == Parameters:
+#
+# [*ensure*]
+#   Whether to install or uninstall an instance.
+#
+# [*version*]
+#   The NodeJS version ('vX.Y.Z', 'latest' or 'stable').
+#
+# [*target_dir*]
+#   Where to install the executables.
+#
+# [*make_install*]
+#   If false, will install from nodejs.org binary distributions.
+#
+# [*cpu_cores*]
+#   Number of CPU cores to use for compiling nodejs. Will be used for parallel 'make' jobs.
+#
+# [*default_node_version*]
+#   The default nodejs version. Required to ensure that the default version won't be uninstalled if $ensure = absent.
+#
+define nodejs::instance($ensure, $version, $target_dir, $make_install, $cpu_cores, $default_node_version) {
+  if $caller_module_name != $module_name {
+    warning('nodejs::instance is private!')
+  }
+
+  validate_string($ensure)
+  validate_integer($cpu_cores)
```

Comment:
yep, but I think that we should still support Puppet 3 (and that's a new feature of puppet4, http://www.example42.com/2015/09/09/puppet4-examples-data-types/). Changing this is an idea for a version 3.